core/src/Threads/Kokkos_ThreadsTeam.hpp

```diff
@@ -888,6 +888,92 @@ Impl::VectorSingleStruct<Impl::ThreadsExecTeamMember> PerThread(
     const Impl::ThreadsExecTeamMember& thread) {
   return Impl::VectorSingleStruct<Impl::ThreadsExecTeamMember>(thread);
 }
+
+template <Kokkos::Iterate Direction, typename... Ns>
+KOKKOS_INLINE_FUNCTION auto MDTeamThreadRange(
+    Impl::ThreadsExecTeamMember const& member, Ns&&... ns) {
+  using execution_space = typename Impl::ThreadsExecTeamMember::execution_space;
+  using array_layout    = typename execution_space::array_layout;
+  static constexpr Kokkos::Iterate outer_direction =
+      Direction == Kokkos::Iterate::Default
+          ? Kokkos::layout_iterate_type_selector<
+                array_layout>::outer_iteration_pattern
+          : Direction;
+  using iType = std::common_type_t<Ns...>;
+
+  return Impl::MDTeamThreadRangeBoundariesStruct<
+      outer_direction, sizeof...(ns), iType, Impl::ThreadsExecTeamMember>(
+      member, static_cast<Ns&&>(ns)...);
+}
+
+template <typename... Ns>
+KOKKOS_INLINE_FUNCTION auto MDTeamThreadRange(
+    Impl::ThreadsExecTeamMember const& member, Ns&&... ns) {
+  return MDTeamThreadRange<Kokkos::Iterate::Default>(member,
+                                                     static_cast<Ns&&>(ns)...);
+}
+
+template <Kokkos::Iterate OuterDirection, Kokkos::Iterate InnerDirection,
+          typename... Ns>
+KOKKOS_INLINE_FUNCTION auto MDThreadVectorRange(
+    Impl::ThreadsExecTeamMember const& member, Ns&&... ns) {
+  using execution_space = typename Impl::ThreadsExecTeamMember::execution_space;
+  using array_layout    = typename execution_space::array_layout;
+  static constexpr Kokkos::Iterate outer_direction =
+      OuterDirection == Kokkos::Iterate::Default
+          ? Kokkos::layout_iterate_type_selector<
+                array_layout>::outer_iteration_pattern
+          : OuterDirection;
+  static constexpr Kokkos::Iterate inner_direction =
+      InnerDirection == Kokkos::Iterate::Default
+          ? Kokkos::layout_iterate_type_selector<
+                array_layout>::outer_iteration_pattern
```

Comment:
Should this rather be `inner_iteration_pattern`? Is `outer_iteration_pattern` a deliberate choice?

Reply:
Good catch. It looks like we made this mistake for all backends. I'll put in a fix for all backends including the serial one.
It looks like our unit-tests fail to catch this bug because all of them were tested with directions specified.

Reply:
Fixed in [fcbc5ee](https://github.com/kokkos/kokkos/pull/4815/commits/fcbc5eec83198fa2dbc48e4961dd05cfb8492222).